ikarus/solver/eigenvaluesolver/generalizedeigensolver.hh

```diff
@@ -0,0 +1,365 @@
+ * \details Under the hood it uses the Eigen::GeneralizedSelfAdjointEigenSolver
+ *
+ * \tparam MT the used Matrix Type, can be a dense Eigen::Matrix
+ */
+template <Concepts::EigenMatrix MT>
+struct GeneralizedSymEigenSolver<EigenValueSolverType::Eigen, MT>
+{
+  using ScalarType = typename MT::Scalar;
+  using MatrixType = MT;
+  using SolverType = Eigen::GeneralizedSelfAdjointEigenSolver<MatrixType>;
+
+  template <typename MATA, typename MATB>
+  requires(Concepts::EigenMatrix<std::remove_cvref_t<MATA>>)
+  GeneralizedSymEigenSolver(MATA&& A, MATB&& B)
+      : matA_(std::forward<MATA>(A)),
+        matB_(std::forward<MATB>(B)),
+        solver_(A.size()) {
+    if ((A.cols() != B.cols()) or (A.rows() != B.rows()) or (A.cols() != B.cols()))
+      DUNE_THROW(Dune::IOError, "GeneralizedSymEigenSolver: The passed matrices should have the same size");
+  }
+
+  template <Concepts::FlatAssembler AssemblerA, Concepts::FlatAssembler AssemblerB>
+  GeneralizedSymEigenSolver(const std::shared_ptr<AssemblerA> assemblerA, const std::shared_ptr<AssemblerB>& assemblerB)
+      : GeneralizedSymEigenSolver(assemblerA->matrix(), assemblerB->matrix()) {}
+
+  /**
+   * \brief Starts the computation of the eigenvalue solver
+   *
+   * \param options defaults to Eigen::ComputeEigenvectors, can be set to Eigen::EigenvaluesOnly, accessing eigenvectors
+   * in that case will results in an error
+   * \return true solving was successful
+   * \return false solving was not successful
+   */
+  bool compute(int options = Eigen::ComputeEigenvectors) {
+    solver_.compute(matA_, matB_, options);
+    computed_ = true;
+    return true; // SelfAdjointEigenSolver will always be successful if prerequisites are met
+  }
+
+  /**
+   * \brief Returns the eigenvalues of the generalized eigenvalue problem
+   *
+   * \return Reference to the vector of eigenvalues
+   */
+  auto& eigenvalues() const {
+    assertCompute();
+    return solver_.eigenvalues();
+  }
+
+  /**
+   * \brief Returns the eigenvectors of the generalized eigenvalue problem
+   *
+   * \param _nev optionally specify how many eigenvectors are requested
+   * \return Reference to the matrix with the eigevectors as columns
+   */
+  auto& eigenvectors() const {
+    assertCompute();
+    return solver_.eigenvectors();
+  }
+
+private:
+  MatrixType matA_;
+  MatrixType matB_;
+  SolverType solver_;
+  bool computed_{};
+
+  void assertCompute() const {
+    if (not computed_)
+      DUNE_THROW(Dune::IOError, "Eigenvalues and -vectors not yet computed, please call compute() first");
+  }
+};
+
+template <EigenValueSolverType tag, Concepts::FlatAssembler AS1, Concepts::FlatAssembler AS2>
+requires(std::same_as<typename AS1::MatrixType, typename AS2::MatrixType> &&
+         not(tag == EigenValueSolverType::Eigen && Concepts::SparseEigenMatrix<typename AS1::MatrixType>))
+auto makeGeneralizedSymEigenSolver(const std::shared_ptr<AS1>& as1, const std::shared_ptr<AS2> as2) {
```

Comment:
> I was thinking of having another helper function, which can be used perhaps also for academic purposes, that solves the problem A x = λ x , assuming B = I , where I is the identity matrix. Can you include it?

This can then also be used in our tests, for example [here](https://github.com/ikarus-project/ikarus/blob/f1928da1cbeaa345c6bddd054afd65b1099a5404/tests/src/testnonlinearelasticity.hh#L273) and [here](https://github.com/ikarus-project/ikarus/blob/f1928da1cbeaa345c6bddd054afd65b1099a5404/tests/src/testtruss.cpp#L241).